src/catalog/src/persist.rs

```diff
@@ -0,0 +1,1079 @@
+};
+use crate::transaction::{
+    add_new_builtin_cluster_replicas_migration, add_new_builtin_clusters_migration,
+    TransactionBatch,
+};
+use crate::{
+    initialize, BootstrapArgs, CatalogError, Cluster, ClusterReplica, Comment, Database,
+    DefaultPrivilege, DurableCatalogError, DurableCatalogState, Epoch, OpenableDurableCatalogState,
+    ReadOnlyDurableCatalogState, ReplicaConfig, Role, Schema, SystemConfiguration,
+    SystemObjectMapping, TimelineTimestamp, Transaction,
+};
+use async_trait::async_trait;
+use differential_dataflow::lattice::Lattice;
+use itertools::Itertools;
+use mz_audit_log::{VersionedEvent, VersionedStorageUsage};
+use mz_controller_types::{ClusterId, ReplicaId};
+use mz_ore::now::NowFn;
+use mz_ore::{soft_assert, soft_assert_eq};
+use mz_persist_client::read::ReadHandle;
+use mz_persist_client::write::WriteHandle;
+use mz_persist_client::{Diagnostics, PersistClient, ShardId};
+use mz_persist_types::codec_impls::{
+    SimpleDecoder, SimpleEncoder, SimpleSchema, UnitSchema, VecU8Schema,
+};
+use mz_persist_types::dyn_struct::{ColumnsMut, ColumnsRef, DynStructCfg};
+use mz_persist_types::Codec;
+use mz_proto::RustType;
+use mz_repr::adt::mz_acl_item::MzAclItem;
+use mz_repr::role_id::RoleId;
+use mz_repr::{Diff, GlobalId};
+use mz_stash::USER_VERSION_KEY;
+use mz_stash_types::objects::proto;
+use mz_stash_types::STASH_VERSION;
+use mz_storage_types::sources::Timeline;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use std::time::Duration;
+use timely::progress::Antichain;
+use uuid::Uuid;
+
+/// New-type used to represent timestamps in persist.
+type Timestamp = u64;
+
+/// Durable catalog mode that dictates the effect of mutable operations.
+#[derive(Debug)]
+enum Mode {
+    /// Mutable operations are prohibited.
+    Readonly,
+    /// Mutable operations have an effect in-memory, but aren't persisted durably.
+    Savepoint,
+    /// Mutable operations have an effect in-memory and durably.
+    Writable,
+}
+
+/// A single update to the catalog state.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct StateUpdate {
+    /// They kind and contents of the state update.
+    kind: StateUpdateKind,
+    /// The timestamp at which the update occurred.
+    ts: Timestamp,
+    /// Record count difference for the update.
+    diff: Diff,
+}
+
+impl StateUpdate {
+    /// Convert a [`TransactionBatch`] to a list of [`StateUpdate`]s at timestamp `ts`.
+    fn from_txn_batch(txn_batch: TransactionBatch, ts: Timestamp) -> Vec<StateUpdate> {
+        fn from_batch<K, V>(
+            batch: Vec<(K, V, Diff)>,
+            ts: Timestamp,
+            kind: fn(K, V) -> StateUpdateKind,
+        ) -> impl Iterator<Item = StateUpdate> {
+            batch.into_iter().map(move |(k, v, diff)| StateUpdate {
+                kind: kind(k, v),
+                ts,
+                diff,
+            })
+        }
```

Comment:
Consider de-structuring TransactionBatch here so if any fields are added to it, we get a compile error until this is updated. Something like:

```
let TransactionBatch{databases, schemas, ...} = txn_batch;
let databases = from_batch(databases, ts, StateUpdateKind::Database);
...
```

Reply:
Great idea, done.